pkg/collector/clock.go

```diff
@@ -140,18 +184,41 @@ func (c *fakeClock) Step(d time.Duration) {
 		// Collect timer functions to run and remove them from list.
 		newTimers := make([]*fakeTimer, 0, len(c.timers))
 		for _, t := range c.timers {
-			if !t.targetTime.After(c.now) {
-				timerFuncs = append(timerFuncs, t.f)
-			} else {
+			if t.targetTime.After(c.now) {
 				newTimers = append(newTimers, t)
+			} else {
+				expiredTimers = append(expiredTimers, t)
 			}
 		}
 		c.timers = newTimers
 	}()
-	// Run the timer functions, without holding a lock. This allows these functions to call
-	// clock.Now(), but also timer.Stop().
-	for _, f := range timerFuncs {
-		f()
+	for _, t := range expiredTimers {
+		if t.f != nil {
+			// Run the timer function, without holding a lock. This allows these
+			// functions to call clock.Now(), but also timer.Stop().
+			t.f()
+		} else {
+			retry := func() bool {
+				if !c.m.TryRLock() {
+					return true
+				}
+				defer c.m.RUnlock()
+				// If timer has been stopped or reset, do not fire.
+				if t.targetTime.IsZero() || t.targetTime.After(c.now) {
+					return false
+				}
+				select {
+				case t.ch <- c.now:
+					return false
+				default:
+				}
+				return true
+			}
+			// Spin until we can acquire a read lock and write to the C channel: this
```

Comment:
Just out of curiosity, why do we only consider the concurrent calls when t.f is nil?

Reply:
The behavior is different for AfterFun-based timers (`t.f != nil`) and channel-based timers (`t.f == nil`). For details, refer to https://pkg.go.dev/time#Timer.Stop. The specified behavior for channel-based timers (Go >= 1.23) is harder to achieve, and the code is more complex.